File: endpoint.js
require('dotenv').config();
const express = require('express');
const path = require('path');
const fs = require('fs');
const pdfjsLib = require('pdfjs-dist');
const { GoogleGenerativeAI } = require("@google/generative-ai");

const app = express();
const PORT = 3000;

// Middleware to parse JSON bodies
app.use(express.json());

// Google Generative AI Setup
const genAI = new GoogleGenerativeAI("AIzaSyAqxVIKyLPplBCNp_QLXPF0T1gAmuh7wbk");
const model = genAI.getGenerativeModel({ model: "gemini-pro" });

// PDF Text Extraction
async function extractTextFromPDF(pdfPath) {
    const data = new Uint8Array(fs.readFileSync(pdfPath));
    const doc = await pdfjsLib.getDocument(data).promise;
    let textContent = '';

    for (let i = 1; i <= doc.numPages; i++) {
        const page = await doc.getPage(i);
        const content = await page.getTextContent();
        textContent += content.items.map(item => item.str).join(' ');
    }

    return textContent;
}

// GET /pdf-text Endpoint
app.get('/pdf-text', async (req, res) => {
    try {
        const pdfText = await extractTextFromPDF('testing.pdf');
        res.json({ pdf_text: pdfText });
    } catch (error) {
        console.error('Error extracting PDF text:', error);
        res.status(500).send('Internal Server Error');
    }
});

// POST /chat Endpoint
app.post('/chat', async (req, res) => {
    try {
        const userQuestion = req.body.user_question;
        const pdfText = await extractTextFromPDF('testing.pdf');
        const prompt = `${pdfText} Use above text and give the following question answer:  ${userQuestion}`;
        const result = await model.generateContent(prompt);
        const response = await result.response;
        res.json({ response: response.text() });
    } catch (error) {
        console.error('Error processing user question:', error);
        res.status(500).send('Internal Server Error');
    }
});

// Serve the index.html file
app.get('/', (req, res) => {
    res.sendFile(path.join(__dirname, 'public', 'index.html'));
});

// Start the server
app.listen(PORT, () => {
    console.log(`Server is running on port ${PORT}`);
});

File: endpoint.py
from flask import Flask, request, jsonify, render_template
import PyPDF2
import google.generativeai as genai

app = Flask(__name__)

def extract_text_from_pdf(pdf_file):
    """
    Function to extract all text from a PDF file.
    
    Args:
    pdf_file (str): The name of the PDF file to extract text from.
    
    Returns:
    str: Text extracted from the PDF.
    """
    text = ""
    with open(pdf_file, 'rb') as file:
        reader = PyPDF2.PdfReader(file)
        for page_num in range(len(reader.pages)):
            text += reader.pages[page_num].extract_text()
    return text

genai.configure(api_key="AIzaSyAqxVIKyLPplBCNp_QLXPF0T1gAmuh7wbk")
# genai.configure(api_key="API KEY")

# Set up the model
generation_config = {
  "temperature": 0.9,
  "top_p": 1,
  "top_k": 1,
  "max_output_tokens": 2048,
}

safety_settings = [
  {
    "category": "HARM_CATEGORY_HARASSMENT",
    "threshold": "BLOCK_MEDIUM_AND_ABOVE"
  },
  {
    "category": "HARM_CATEGORY_HATE_SPEECH",
    "threshold": "BLOCK_MEDIUM_AND_ABOVE"
  },
  {
    "category": "HARM_CATEGORY_SEXUALLY_EXPLICIT",
    "threshold": "BLOCK_MEDIUM_AND_ABOVE"
  },
  {
    "category": "HARM_CATEGORY_DANGEROUS_CONTENT",
    "threshold": "BLOCK_MEDIUM_AND_ABOVE"
  },
]

model = genai.GenerativeModel(model_name="gemini-1.0-pro",
                              generation_config=generation_config,
                              safety_settings=safety_settings)

convo = model.start_chat(history=[])

@app.route('/', methods=['GET', 'POST'])
def chat():
    if request.method == 'POST':
        user_question = request.form['user_question']
        response = convo.send_message(pdf_text + " Use above text and give the following question answer: " + user_question)
        return jsonify({'response': response.text})
    return render_template('index.html')
@app.route('/pdf-text', methods=['GET'])
def get_pdf_text():
    pdf_file = 'testing.pdf'  # Replace 'testing.pdf' with the name of your PDF file
    pdf_text = extract_text_from_pdf(pdf_file)
    return jsonify({'pdf_text': pdf_text})
  
if __name__ == "__main__":
    pdf_file = 'testing.pdf'  # Replace 'testing.pdf' with the name of your PDF file
    pdf_text = extract_text_from_pdf(pdf_file)
    app.run(debug=True)
